executor/src/wasm_utils.rs

```diff
@@ -0,0 +1,132 @@
+	() => ([]);
+	( $( $t:ty ),* ) => ( [ $( { use $crate::wasm_utils::ConvertibleToWasm; <$t>::VALUE_TYPE }, )* ] );
+}
+
+#[macro_export]
+macro_rules! convert_fn {
+	( $name:ident ( $( $params:ty ),* ) ) => ( $crate::wasm_utils::UserFunctionDescriptor::Static(stringify!($name), &convert_args!($($params),*), None) );
+	( $name:ident ( $( $params:ty ),* ) -> $returns:ty ) => ( $crate::wasm_utils::UserFunctionDescriptor::Static(stringify!($name), &convert_args!($($params),*), Some({ use $crate::wasm_utils::ConvertibleToWasm; <$returns>::VALUE_TYPE }) ) );
+}
+
+#[macro_export]
+macro_rules! reverse_params {
+	// Entry point, use brackets to recursively reverse above.
+	($body:tt, $self:ident, $context:ident, $( $names:ident : $params:ty ),*) => (
+		reverse_params!($body $self $context [ $( $names : $params ),* ]);
+	);
+	($body:tt $self:ident $context:ident [] $( $names:ident : $params:ty ),*) => ({
+		$(
+			let $names : <$params as $crate::wasm_utils::ConvertibleToWasm>::NativeType = match $context.value_stack.pop_as() {
+				Ok(value) => value,
+				Err(error) => return Err(error.into()),
+			};
+		)*
+		$body
+	});
+	($body:tt $self:ident $context:ident [ $name:ident : $param:ty $(, $names:ident : $params:ty )* ] $( $reversed_names:ident : $reversed_params:ty ),*) => (
+		reverse_params!($body $self $context [ $( $names : $params ),* ] $name : $param $( , $reversed_names : $reversed_params )*);
+	);
+}
+
+#[macro_export]
+macro_rules! marshall {
+	( $context:ident, $self:ident, ( $( $names:ident : $params:ty ),* ) -> $returns:ty => $body:tt ) => ({
+		let r : <$returns as $crate::wasm_utils::ConvertibleToWasm>::NativeType = reverse_params!($body, $self, $context, $( $names : $params ),*);
+		Ok(Some({ use $crate::wasm_utils::ConvertibleToWasm; r.to_runtime_value() }))
+	});
+	( $context:ident, $self:ident, ( $( $names:ident : $params:ty ),* ) => $body:tt ) => ({
+		reverse_params!($body, $self, $context, $( $names : $params ),*);
+		Ok(None)
+	})
+}
+
+#[macro_export]
+macro_rules! dispatch {
+	( $objectname:ident, $( $name:ident ( $( $names:ident : $params:ty ),* ) $( -> $returns:ty )* => $body:tt ),* ) => (
+		fn execute(&mut self, name: &str, context: $crate::wasm_utils::CallerContext)
+			-> $crate::wasm_utils::result::Result<Option<$crate::wasm_utils::RuntimeValue>, $crate::wasm_utils::Error> {
+			let $objectname = self;
+			match name {
+				$(
+					stringify!($name) => marshall!(context, $objectname, ( $( $names : $params ),* ) $( -> $returns )* => $body),
+				)*
+				n => Err($crate::wasm_utils::Error::Trap(format!("not implemented: {}", n)).into())
+			}
+		}
+	);
+}
+
+#[macro_export]
+macro_rules! signatures {
+	( $( $name:ident ( $( $params:ty ),* ) $( -> $returns:ty )* ),* ) => (
+		const SIGNATURES: &'static [$crate::wasm_utils::UserFunctionDescriptor] = &[
+			$(
+				convert_fn!( $name ( $( $params ),* ) $( -> $returns )* ),
+			)*
+		];
+	);
+}
+
+#[macro_export]
+macro_rules! function_executor {
```

Comment:
could rename to something like `impl_function_executor`. also could use a doc string.
I'd recommend either moving the other macros into other arms of the same macro, with e.g. `signatures!` becoming a branch 

`( SIGNATURES $( $name:ident ( $( $params:ty ),* ) $( -> $returns:ty )* ),*)`